### ISSUE TITLE
Advanced Audit tests flaking

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#60719]
Started flaking yesterday: https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=Advanced%20Audit

Seems to coincide with https://github.com/kubernetes/kubernetes/pull/60237

/assign @tallclair @crassirostris 

need to triage this ASAP to know if we need to roll back that PR

**Comment by BenTheElder** [comment]
/milestone v1.10
(should not work, lack of perms)

**Comment by tallclair** [comment]
From [apiserver log](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/22334/artifacts/bootstrap-e2e-master/kube-apiserver.log) for run [ci-kubernetes-e2e-gci-gce/22334](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/22334):
```
I0302 17:08:26.869897       1 flags.go:27] FLAG: --audit-log-batch-buffer-size="10000"
I0302 17:08:26.869904       1 flags.go:27] FLAG: --audit-log-batch-max-size="400"
I0302 17:08:26.869910       1 flags.go:27] FLAG: --audit-log-batch-max-wait="30s"
I0302 17:08:26.869919       1 flags.go:27] FLAG: --audit-log-batch-throttle-burst="15"
I0302 17:08:26.869925       1 flags.go:27] FLAG: --audit-log-batch-throttle-enable="true"
I0302 17:08:26.869931       1 flags.go:27] FLAG: --audit-log-batch-throttle-qps="10"
I0302 17:08:26.869941       1 flags.go:27] FLAG: --audit-log-format="json"
I0302 17:08:26.869947       1 flags.go:27] FLAG: --audit-log-maxage="0"
I0302 17:08:26.869953       1 flags.go:27] FLAG: --audit-log-maxbackup="0"
I0302 17:08:26.869960       1 flags.go:27] FLAG: --audit-log-maxsize="2000000000"
I0302 17:08:26.869966       1 flags.go:27] FLAG: --audit-log-mode="batch"
I0302 17:08:26.869972       1 flags.go:27] FLAG: --audit-log-path="/var/log/kube-apiserver-audit.log"
I0302 17:08:26.869979       1 flags.go:27] FLAG: --audit-policy-file="/etc/audit_policy.config"
I0302 17:08:26.869986       1 flags.go:27] FLAG: --audit-webhook-batch-buffer-size="10000"
I0302 17:08:26.869992       1 flags.go:27] FLAG: --audit-webhook-batch-initial-backoff="10s"
I0302 17:08:26.869999       1 flags.go:27] FLAG: --audit-webhook-batch-max-size="400"
I0302 17:08:26.870005       1 flags.go:27] FLAG: --audit-webhook-batch-max-wait="30s"
I0302 17:08:26.870012       1 flags.go:27] FLAG: --audit-webhook-batch-throttle-burst="15"
I0302 17:08:26.870018       1 flags.go:27] FLAG: --audit-webhook-batch-throttle-enable="false"
I0302 17:08:26.870024       1 flags.go:27] FLAG: --audit-webhook-batch-throttle-qps="10"
I0302 17:08:26.870031       1 flags.go:27] FLAG: --audit-webhook-config-file=""
I0302 17:08:26.870037       1 flags.go:27] FLAG: --audit-webhook-initial-backoff="10s"
I0302 17:08:26.870044       1 flags.go:27] FLAG: --audit-webhook-mode="batch"
```

I'm pretty sure this is what's happening: the default audit mode changed to batch for the logging backend, but the audit test expects the logs to be there immediately.

**Recommended Action:**

1. Revert the defaults on the audit flags. I don't think we can change the default behavior for backwards compatibility.
2. Add retry logic to the audit test so that it can handle async logs.
3. (optional) update the e2e configuration to use batch mode with a very short timeout.

If we rollback #60237 it will cause problems with the scale tests that were re-enabled in https://github.com/kubernetes/test-infra/pull/7000

**Comment by krzyzacy** [comment]
http://k8s-testgrid.appspot.com/sig-release-master-blocking#gci-gce
starts to look pretty bad
priority/critical-urgent
cc @jberkus

**Comment by tallclair** [comment]
Actually, the forward fix would break scalability anyway. I'll send a rollback.

**Comment by caoshuilanting** [comment]
sdfs

**Comment by crassirostris** [comment]
@tallclair 

> Actually, the forward fix would break scalability anyway. I'll send a rollback.

No, because audit logging _tests_ are not enabled in scalability tests, there's [`DisabledForLargeClusters` in the test name](https://github.com/kubernetes/kubernetes/blob/baf93c060e04ca9caf8a4a648d39b58eb7181449/test/e2e/auth/audit.go#L60)

**Comment by crassirostris** [comment]
I think the e2e test fix is better in this case

**Comment by crassirostris** [comment]
Filed https://github.com/kubernetes/kubernetes/pull/60794 to fix the problem

I still believe that using buffered audit logging is a better default, since slowing down all api requests is a pretty serious drawback and should be enabled consciously

**Comment by crassirostris** [comment]
@liggitt @tallclair If you're fine with forward-fixing the problem in e2es, I'd prefer this way, it's the easiest. Fix is in https://github.com/kubernetes/kubernetes/pull/60794

**Comment by tallclair** [comment]
> I still believe that using buffered audit logging is a better default, since slowing down all api requests is a pretty serious drawback and should be enabled consciously

If we make batch mode the default, I think we should adjust the default parameters to include:

1. Very low timeout (1s?) - I don't think there's any reason to collect large batches.
2. Only a single go-routine (do we have the right params for this?) - This is important to reduce write contention, and to keep log lines in order.

I'm OK moving forward with the test fix, but I'd like to fix those defaults before 1.10 is cut, if possible.

**Comment by crassirostris** [comment]
> I'm OK moving forward with the test fix, but I'd like to fix those defaults before 1.10 is cut, if possible.

I agree that defaults should change. What about changing the batch size to 1, so that each log entry is written asynchronously in its own goroutine? Anyways, this or very low timeout SGTM, since I don't think that the second option is feasible right now.

**Comment by liggitt** [comment]
still failing in upgrade test... e2e fix likely needs backport to 1.9 e2es
http://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-1.9-master-upgrade-master

**Comment by crassirostris** [comment]
Good point! Created https://github.com/kubernetes/kubernetes/pull/61134 to address it

**Comment by liggitt** [comment]
fixed in upgrade test by #61134
